eventbridge/eventbridge.go

```diff
@@ -0,0 +1,173 @@
+
+// NewGetterService returns a server-side implementation of the Getter interface. Given an existing event source, it
+// will subscribe to that source with no filters. Each GetEvent call pops and returns an event buffered in the
+// subscription's channel of published events. If no event is buffered, it blocks until one is.
+func NewGetterService(ctx context.Context, eventSource events.Subscriber) Getter {
+	eventChan, errChan := eventSource.Subscribe(ctx)
+	return &getterService{
+		eventChan: eventChan,
+		errChan:   errChan,
+	}
+}
+
+// GetEvent pops and returns an event buffered in the service's subscription channel.
+func (s *getterService) GetEvent(ctx context.Context) (*eventapi.Envelope, error) {
+	select {
+	case receivedEnvelope := <-s.eventChan:
+		return &eventapi.Envelope{
+			Timestamp: receivedEnvelope.Timestamp,
+			Namespace: receivedEnvelope.Namespace,
+			Topic:     receivedEnvelope.Topic,
+			Event:     receivedEnvelope.Event,
+		}, nil
+	case err := <-s.errChan:
+		return nil, err
+	case <-ctx.Done():
+		if ctxErr := ctx.Err(); ctxErr != context.Canceled {
+			return nil, ctxErr
+		}
+		return nil, nil
+	}
+}
+
+// RegisterGetterService adds the Getter service as a method to the provided TTRPC server.
+func RegisterGetterService(srv *ttrpc.Server, svc Getter) {
+	srv.Register(getterServiceName, map[string]ttrpc.Method{
+		getEventMethodName: func(ctx context.Context, unmarshal func(interface{}) error) (interface{}, error) {
+			return svc.GetEvent(ctx)
+		},
+	})
+}
+
+type getterClient struct {
+	rpcClient *ttrpc.Client
+}
+
+// NewGetterClient returns a client-side implementation of the Getter interface using the provided TTRPC client to
+// connect to the server.
+func NewGetterClient(rpcClient *ttrpc.Client) Getter {
+	return &getterClient{
+		rpcClient: rpcClient,
+	}
+}
+
+// GetEvent requests and returns an event from the Getter service. If no event is available at the time of the request
+// it will block until one is.
+func (c *getterClient) GetEvent(ctx context.Context) (*eventapi.Envelope, error) {
+	var req types.Empty
+	var resp eventapi.Envelope
+	if err := c.rpcClient.Call(ctx, getterServiceName, getEventMethodName, &req, &resp); err != nil {
+		return nil, err
+	}
+
+	return &resp, nil
+}
+
+// Attach takes an existing Getter and forwards each event retrieved from it to the provided forwarder. For example, if
+// the source is a client for a remote exchange and sink is a local exchange, Attach will result in each event published
+// on the remote exchange to be forwarded to the local exchange, essentially bridging the two together.
+func Attach(ctx context.Context, source Getter, sink events.Forwarder) error {
+	for {
+		// If the context was closed, return. Otherwise, don't get hung up and continue on with the loop.
+		select {
+		case <-ctx.Done():
+			if ctxErr := ctx.Err(); ctxErr != context.Canceled {
+				return ctxErr
+			}
+			return nil
+		default:
+		}
+
+		envelope, err := source.GetEvent(ctx)
```

Comment:
I have a concern with this loop just spinning since it makes the assumption that `GetEvent` will handle not spinning this loop. I think another method of `HasEvent` or something of that nature needs to be added and should behave similarly to how `ctx.Done` works by returning a channel.

Reply:
The Godoc for getterClient.GetEvent mentions that it's blocking, I can add an additional short comment above the line your commenting on re-iterating it.

Reply:
I think this needs to be added to the interface's method signature of `GetEvent`. There is no mention of that in the interface.

Reply:
Yeah I just noticed that too, I am adding a comment to the interface method too in addition to the implementations.